OSOE-111: Detect visual errors (visual verification testing) in the UI Testing Toolbox

dministro opened the pull request:
OSOE-111

dministro committed:
Updating submodules:
 - Lombiq.ChartJs: changing to extensions moved to Lombiq.UITestingToolbox
 - Lombiq.UITestingToolbox: adding visual error detection related extensions

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodules:
 - Lombiq.ChartJs: merge
 - Lombiq.UITestingToolbox: merge

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodules:
 - Lombiq.ChartJs: merge
 - Lombiq.UITestingToolbox: merge

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: chrome color profile

dministro committed:
Updating nuget package versions
Updating submodules:
 - Lombiq.ChartJs: package version, assembly resource
 - Lombiq.Privacy: package version
 - Lombiq.UITestingToolbox: basic visual test

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating nuget package versions
updating submodules:
 - Lombiq.ChartJs: package version
 - Lombiq.Privacy: package version
 - Lombiq.UITestingToolbox: crop

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: reference image, threshold

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: threshold

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: roi, threshold

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: roi

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: threshold

dministro committed:
updating nuget package versions
updating submodules:
 - Lombiq.ChartJs: package version
 - Lombiq.Privacy: package version

dministro committed:
temporary upload Temp folder in github action

dministro committed:
fix workflow

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: dpi scaling

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: viewport size

dministro committed:
set screen resoulution to fhd on windows under tests

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: cropping, file paths

dministro committed:
removing upload temp folder from GH workflow
updating submodules:
 - Lombiq.ChartJs: nuget package version
 - Lombiq.Privacy: nuget package version

Piedone commented on .github/workflows/build-and-test.yml:
```diff
@@ -124,6 +124,7 @@ jobs:
             export Lombiq_Tests_UI__DockerConfiguration__ContainerName="sql2019"
         else
             export Lombiq_Tests_UI__SqlServerDatabaseConfiguration__ConnectionStringTemplate="Server=.\\SQLEXPRESS;Database=LombiqUITestingToolbox_{{id}};Integrated Security=True;MultipleActiveResultSets=True;Connection Timeout=60;ConnectRetryCount=15;ConnectRetryInterval=5"
+            powershell Set-DisplayResolution -Width 1920 -Height 1080 -Force
```
This shouldn't be necessary as UI testing shouldn't rely on the screen resolution. That's why we also have `SetBrowserSize()`.

dministro replied:
Unfortunately, this or something similar is necessary, because, without this, the browser window size is smaller than expected on windows-latest.
Below you can find screenshots and sizes before and after setting screen resolution on windows-latest.

**ubuntu-latest:**
size: 1600 ×900
![ubuntu-latest_navbar_full_screen](https://user-images.githubusercontent.com/15349881/172018490-c2d57bcb-0416-4297-93b3-4bdd5e6d673c.png)

**windows-latest**-before setting screen resolution
size: 1028 × 655
![windows-latest_navbar_full_screen](https://user-images.githubusercontent.com/15349881/172018492-7a1d504c-37e9-4490-8b6e-9440cc4b2fd8.png)

**windows-latest**-after setting screen resolution
size: 1600 ×900
![windows-latest_after_set_screen_size_navbar_full_screen](https://user-images.githubusercontent.com/15349881/172018491-c723841a-847f-4523-b9ec-ee0150400a9d.png)

Piedone replied:
Please look into how this works with GitHub Actions, and why this difference between platforms is there.

dministro replied:
**Maximum window size on windows-latest:**
If I remember well, the maximum window size can't be larger than the desktop size(plus some pixels) on windows.
I didn't find official documentation about this, but I tested it in non-headless mode:
My desktop size: 3840×1200
Requested window size: context.SetBrowserSize(CommonDisplayResolutions.FiveK)
The resulted window size: 3860×1220
![image](https://user-images.githubusercontent.com/15349881/172785146-b2f6e5ed-1c7b-4f73-b283-cd10889489a4.png)

**GH windows-latest default screen resolution:**
I didn't find official documentation about this, but here is an issue(related to Windows 2019 and screen size): https://github.com/actions/virtual-environments/issues/2935

I'll give it a try to run UI tests in headless mode on windows-latest. As I see in the workflow and configuration they are started in non-headless mode.

dministro replied:
It's not surprising, but all tests passed on windows-latest in headless mode without setting screen resolution.

So, which mode is preferred on windows-latest? headless or non-headless?

Piedone replied:
Great then! Headless all the way. However, how was it not headless before? Hmm, `UITestBase` makes it default to `false` in both test projects (due to the global default being `true`). I think we did that to aid in inspecting them locally, due to the demonstration purpose the two test projects serve. So, you can just set that config as `true` in the GitHub build's `Tests` step. (In TeamCity it's set as such from its config.)

Piedone reviewed (changes requested):
Merge from dev, please.

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: comments, readme

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: merge

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.ChartJs: extension method name

dministro committed:
updating submodules:
 - Lombiq.ChartJs: ui test
 - Lombiq.DataTables: merge
 - Lombiq.UITestingToolbox: ShouldVisualMatchApproved, failure dump

dministro committed:
updating submodules:
 - Lombiq.ChartJs: reference image as resource
 - Lombiq.UITestingToolbox: reference image as resource, loading reference images from embedded resource

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: bump Codeuctivity.ImageSharpCompare version to 2.0.46

dministro committed:
updating nuget package versions
bump OrchardCore nuget package versions
updating submodules:
 - Lombiq.ChartJs: nuget package version
 - Lombiq.Privacy: nuget package version

dministro committed:
updating nuget package versions
updating submodules:
 - Lombiq.HelpfulLibraries: package version
 - Lombiq.DataTables: package version
 - Lombiq.HelpfulExtensions: package version
 - Lombiq.Hosting.Azure.ApplicationInsights: merge
 - Lombiq.Hosting.Tenants: merge
 - Lombiq.JsonEditor: merge
 - Lombiq.TrainingDemo: merge
 - Lombiq.VueJs: merge
 - Lombiq.BaseTheme: package version

dministro committed:
removing quick fix

dministro committed:
fixing test issues after updating OC1.4

dministro committed:
updating submodules:
 - Lombiq.ChartJs: extension name changed
 - Lombiq.UITestingToolbox: friendly assertion message

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating nuget package versions
updating submodules:
 - Lombiq.ChartJs: nuget package version

dministro committed:
Update build-and-test.yml

run ui tests in headless mode

dministro committed:
Update build-and-test.yml

changing to headless mode for windows and linux tests

Piedone commented on NuGetTest/test/Lombiq.OSOCE.NuGet.Tests.UI/Lombiq.OSOCE.NuGet.Tests.UI.csproj:
```diff
@@ -11,16 +11,16 @@
 
   <ItemGroup>
     <Content Include=".htmlvalidate.json">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
```
Why is this necessary?

dministro replied:
This was an accident.

This change was because there are 3 .htmlvalidate.json under OSOCE and all of it is set to `PreserveNewest`. One of them is in Lombiq.Tests.UI NuGet package, is a dependency of Lombiq.OSOCE.NuGet.Tests.UI project, which contains .htmlvalidate.json too. After I created a new prerelease from Lombiq.Tests.UI project, it become newer, and this caused problems on the local run of NuGet tests.

Piedone commented on NuGetTest/src/Themes/Lombiq.OSOCE.NuGet.TestTheme/Lombiq.OSOCE.NuGet.TestTheme.csproj:
```diff
@@ -25,10 +25,10 @@
-    <PackageReference Include="OrchardCore.ContentManagement" Version="1.3.0" />
-    <PackageReference Include="OrchardCore.DisplayManagement" Version="1.3.0" />
-    <PackageReference Include="OrchardCore.ResourceManagement" Version="1.3.0" />
+    <PackageReference Include="OrchardCore.Theme.Targets" Version="1.4.0" />
```
Thanks for noticing these!

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
changing to headless mode for windows and linux tests

dministro committed:
Merge branch 'issue/OSOE-111' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodules:
 - Lombiq.ChartJs: ui test
 - Lombiq.UITestingToolbox: comments, names, extensions, By suffix

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: element screenshot, full page UI test

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: Failure dump exception message

dministro committed:
setting MaxParallelTests=1

dministro committed:
updating submodules:
 - Lombiq.HelpfulLibraries: merge
 - Lombiq.ChartJs: merge
 - Lombiq.DataTables: merge
 - Lombiq.HelpfulExtensions: merge
 - Lombiq.Privacy: merge
 - Lombiq.BaseTheme: merge
 - Lombiq.UITestingToolbox: merge

dministro committed:
setting MaxParallelTests=1 only for Lombiq.Tests.UI.Samples

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
enable only VerifyHomePage

dministro committed:
fixing run only VerifyHomePage from Lombiq.Tests.UI.Samples only on ubuntu-latest

dministro committed:
disabling NuGetTests

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: bitmap version

dministro committed:
restoring GH workflow
restoring .htmlvalidate.json settings in Lombiq.OSOCE.NuGet.Tests.UI.csproj
updating submodules:
 - Lombiq.Gulp.Extensions: merge
 - Lombiq.Npm.Targets: merge
 - Lombiq.SetupExtensions: merge
 - Lombiq.Tests: merge
 - Lombiq.UITestingToolbox: tuning UI test
 - Lombiq.Analyzers: merge

dministro committed:
restoring GH workflow again

dministro committed:
restoring Lombiq.OSOCE.NuGet.Tests.UI.csproj again

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: configuration

Piedone commented:
Merge from dev, please.

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodules:
 - Lombiq.HelpfulLibraries: adding JoinNotNullOrEmpty
 - Lombiq.UITestingToolbox: fixes

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: hide/restore scrollbar

dministro committed:
updating submodules:
 - Lombiq.ChartJs: fine tuning
 - Lombiq.UITestingToolbox: full page screenshot, fine tuning

Piedone commented on .github/workflows/build-and-test.yml:
```diff
@@ -141,6 +141,11 @@ jobs:
             }
         }
+        
+        Configure-Environment @{
+            Path = 'Env:\Lombiq_Tests_UI__BrowserConfiguration__Headless'
+            Value = 'true'
+        }
```
You'll need to add this to https://github.com/Lombiq/GitHub-Actions/blob/dev/.github/actions/test-dotnet/run-tests.ps1 after merging from dev. And there it would be simpler to just write `$Env:Lombiq_Tests_UI__BrowserConfiguration__Headless = "true"` instead, and changing the existing `Configure-Environment`s to that too.

dministro replied:
Thank you for the info.

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
changing GH actions to issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: failure dumps

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: fine tuning WaitScrollToNotChange

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: fine tuning WaitScrollToNotChange

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: fixing error SA1505

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: fine tuning WaitScrollToNotChange

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: tuning WaitScrollToNotChange

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: tuning WaitScrollToNotChange

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: tuning WaitScrollToNotChange

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: tuning WaitScrollToNotChange

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: comments

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: removing RestoreHiddenScrollBar

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: trying to remove Thread.Sleep from the beginning of WaitScrollToNotChange

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: font-family
 - Lombiq.UITestingToolbox: threshold

dministro committed:
updating submodules:
 - Lombiq.VueJs: removing duplicated package reference
 - Lombiq.BaseTheme: open-sans
 - Lombiq.UITestingToolbox: reference images, threshold

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: open-sans again
 - Lombiq.UITestingToolbox: reference image for open-sans

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: setting default font to merriweather
 - Lombiq.UITestingToolbox: reference images for merriweather

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: setting default font to ubuntu
 - Lombiq.UITestingToolbox: reference images for ubuntu

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: setting default font to roboto
 - Lombiq.UITestingToolbox: reference images for roboto

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: setting default font to segoe-ui
 - Lombiq.UITestingToolbox: reference images for segoe-ui

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: open-sans the winner
 - Lombiq.UITestingToolbox: changing font rendering chrome flags

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: more font rendering related chrome flags

dministro committed:
updating submodules:
 - Lombiq.BaseTheme: restoring site font
 - Lombiq.UITestingToolbox: lower quality image comparison

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: removing font rendering related chrome args

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: mean error

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: adding unix reference images, browsername suffix, comments

dministro committed:
updating submodules:
 - Lombiq.DataTables: change to dev, PR closed
 - Lombiq.HelpfulExtensions: change to dev, PR closed
 - Lombiq.Privacy: change to dev, PR closed
 - Lombiq.BaseTheme: change to dev, PR closed

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: "reference" to "baseline"

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: comments

dministro committed:
updating submodule:
 - Lombiq.HelpfulLibraries: merge

dministro committed:
updating submodules:
 - Lombiq.ChartJs: param name
 - Lombiq.UITestingToolbox: comments, failure dump

dministro committed:
merge from dev

dministro committed:
updating submodule:
 - Lombiq.GitHub.Actions: switch to issue/OSOE-111

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: webdriver configuration, tests

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: removing edge annotation, updating firefox baseline image

dministro committed:
updating submodule:
 - adding custom edge config

dministro committed:
updating submodules:
 - Lombiq.GitHub.Actions: setup-edge
 - Lombiq.UITestingToolbox: comments

dministro committed:
updating submodule:
 - Lombiq.GitHub.Actions: fixing typo

dministro committed:
updating submodules:
 - Lombiq.GitHub.Actions: fix
 - Lombiq.UITestingToolbox: adding edge and firefox annotations and baseline images

dministro committed:
updating submodules:
 - Lombiq.GitHub.Actions: setup edge only on linux

dministro committed:
updating submodule:
 - Lombiq.GitHub.Actions: typo

dministro committed:
updating submodule:
 - Lombiq.GitHub.Actions: fixing variable name

dministro committed:
updating submodules:
 - Lombiq.GitHub.Actions: fixing variable name
 - Lombiq.UITestingToolbox: updating baseline images for edge on linux

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: removing firefox annotation and related baseline images

dministro committed:
updating submodules:
 - Lombiq.GitHub.Actions: comments
 - Lombiq.UITestingToolbox: comments

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

Piedone reviewed (changes requested):
Merge from dev, please.

dministro committed:
adding setup-edge-on-linux parameter to wf
updating submodules:
 - Lombiq.GitHub.Actions: conditional setup edge
 - Lombiq.UITestingToolbox: comments

dministro committed:
updating submodules:
 - Lombiq.GitHub.Actions: step position
 - Lombiq.UITestingToolbox: comments

dministro committed:
removing setup edge from nuget tests

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: comments

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: comments

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-111

dministro committed:
updating submodule:
 - Lombiq.GitHub.Actions: docs

dministro committed:
updating submodule:
 - Lombiq.UITestingToolbox: Bitmap to Image in method params

dministro committed:
updating submodule:
 - Lombiq.GitHub.Actions: alphabetical order

Piedone committed:
Updating submodules to merged dev

Piedone committed:
NuGet releases and updating Lombiq packages to latest

Piedone committed:
Separate NuGet UI test for DataTables and ChartJs for parallelization

Piedone committed:
Fixing that ConsentBannerShouldWorkWithRazorAndLiquidBasedThemes failed due to too long page title

Piedone committed:
Updating NuGet solution

Piedone committed:
Revert "Updating NuGet solution"

This reverts commit c078493e6ea2010fb6d1746c13ec192c20f7b6a2.
